File: array/1475_final_prices_with_a_special_discount_in_a_shop/1475_final_prices_with_a_special_discount_in_a_shop.go
package main

import "fmt"

func finalPrices(prices []int) []int {
	ans := make([]int, len(prices))
	var stack []int
	for i, p := range prices {
		fmt.Println(i, p)
		for len(stack) > 0 {
			idx := stack[len(stack)-1]
			if p <= prices[idx] {
				ans[idx] = prices[idx] - p
				stack = stack[0 : len(stack)-1]
			} else {
				break
			}
		}
		stack = append(stack, i)
	}
	fmt.Println(stack)
	for _, idx := range stack {
		ans[idx] = prices[idx]
	}
	return ans
}

func main() {
	prices := []int{8, 4, 6, 2, 3}
	res := finalPrices(prices)
	fmt.Println(res)
}

File: array/1475_final_prices_with_a_special_discount_in_a_shop/1475_final_prices_with_a_special_discount_in_a_shop.py
from typing import List


class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        ans = []
        n = len(prices)
        for i in range(n):
            dis = 0
            for j in range(i+1, n):
                if prices[j]<=prices[i]:
                    dis = prices[j]
                    break
            ans.append(prices[i]-dis)
        return ans


prices = [8,4,6,2,3]
res = Solution().finalPrices(prices)
print(res)